### PR TITLE
Remove Base64 encode and decode methods from Azure::Storage and move the string-based helper to Internal. Update storage call sites.

### LINKED ISSUE #1020: Move Base-64 API from Storage to CORE
•	Base-64 encode/decode functions convert to/from byte[] to string

API is expected to be re-usable from any SDK Service.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Follow up from https://github.com/Azure/azure-sdk-for-cpp/pull/1297 as the second step to fix #1020

### CODE REVIEW

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:7]: storage code has an unwritten rule of the order of these include headers. Usually we put this header after C++ standard library headers

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:16]: What's the motivation of adding the suffix `Text` and moving it into `Internal` namespace? Why cannot this overloaded function be moved to `Azure::Core::` together with the other 2 functions?

File: sdk/storage/azure-storage-blobs/test/block_blob_client_test.cpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-blobs/test/block_blob_client_test.cpp:101]: Should we come into agreement on the version of clang-format we use so as to avoid this kind of changes in the future?

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:16]: This is more of a convenience wrapper around the underling Base64 encode method and we shouldn't expose such things to the customer unless there is a strong reason to. I don't expect other services and client libraries to call or use such an API, and is storage specific.

In this particular case, Base64 is an encoding from binary data to string, and the encode method accepting a string parameter is confusing. Ideally, this internal method would be removed and all the call sites use the `vector<uint8_t>` based overload.

I added `Text` to reconcile this oddity and make it explicit that this isn't the "expected" encode method.

File: sdk/storage/azure-storage-blobs/test/block_blob_client_test.cpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-blobs/test/block_blob_client_test.cpp:101]: I thought we did. Anything 9 or above. I am using `10.0.0`. What version were you using?
https://github.com/Azure/azure-sdk-for-cpp/blob/910731893a2c113392f3be1979045cc9c5847bae/CONTRIBUTING.md#third-party-dependencies

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (vhvb1989) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:14]: Why are we making storage common to wrap the functionality from Azure core? What's the point of it?

The API is already exposed for all services. We shouldn't put it under another API within common

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:7]: basically @katmsft and I want to follow https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes because our guidelines lack such a rule.

I totally understand if you don't like the rule from Google C++ style guide. But before we add a new regulation into our guidelines, can we just make it consistent within storage sdk?

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:14]: This is a convenience overload that happens to accept `std::string` and just an implementation detail to avoid having to re-write all the callers.

The alternative, and correct approach (imo), is to update all callers of `Base64EncodeText` to call the byte vector based public `Base64Encode` API coming from `Azure::Core`.

I will leave it up to the storage folks to remove this, if and when they want to, but it shouldn't be public.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/protocol/blob_rest_client.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/protocol/blob_rest_client.hpp:4346]: Too bad, we shouldn't edit auto-generated code manually.

Can you open an issue and assign it to me to update the auto-generator?

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp]: ok

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp]: Yes, makes sense. Already updated to fix that :)

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/protocol/blob_rest_client.hpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/protocol/blob_rest_client.hpp:4346]: How can I remove the method definition in Azure::Storage then in this PR, when the auto-generated code is still calling it?

Can you update the auto-generator and add a commit to this PR directly?

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:14]: Hi @ahsonkhan I was thinking, how about we add two functions to convert between `std::string` and `std::vector<uint8_t>` in `azure/core/strings.hpp`. Maybe we want to put them into `Internal` namespace because we don't want to expose them to end users, they're just for azure sdk developers. These two functions can be super useful. This way, we don't need `Base64EncodeText` anymore, and the code in `sdk/core/azure-core/test/ut/base64.cpp` can also be greatly simplified.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/protocol/blob_rest_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/protocol/blob_rest_client.hpp:4346]: I think you can just merge this PR with these changes as is now. I'll update auto-generator later.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:14]: That seems reasonable as a subsequent change after the Jan milestone, since it will be non-breaking.

Is it safe to assume that `std::string` in C++ will be UTF-8 encoded bytes and is the reinterpret cast for each string character to `uint8_t` allowed?

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/protocol/blob_rest_client.hpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/protocol/blob_rest_client.hpp:4346]: Sounds good. Filed an issue: https://github.com/Azure/azure-sdk-for-cpp/issues/1299

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:14]: While i'm thinking the implementation will be as simple as

```
std::string ToString(const std::vector<uint8_t>& data) {
    return std::string(data.begin(), data.end());
}

std::vector<uint8_t> ToBinary(const std::string& data) {
    return std::vector<uint8_t>(data.begin(), data.end());
}
```

Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:14]: > While i'm thinking the implementation will be as simple as

That's what I am concerned with.

Where do you expect to use this? Depending on the encoding of the binary data, the string representation could be different. What do you expect converting any arbitrary bytes into `std::string` to mean exactly?

For example, imagine this arbitrary byte vector (which doesn't contain valid UTF-8 characters):
```C++
   vector<uint8_t> vect { 226, 40, 161 }; 
   string someText = ToString(vect); // What is the expectation here?
   cout << "Some output: " << someText << endl; // Some non-printable output: �(�
```

File: sdk/core/azure-core/test/ut/base64.cpp
Comment (vhvb1989) [inline sdk/core/azure-core/test/ut/base64.cpp:111]: out of curiosity, if this ever fail, how would we do reproduce it?  Since it is random generation, How do we know what was the data content?  
Maybe add a print to console to display the data raw data upon failure?  gtest would only display the Encoded and Decoded data right?

**Comment by ahsonkhan** [comment]
@Jinming-Hu, @vhvb1989 - looking for a review and approval :)

File: sdk/core/azure-core/test/ut/base64.cpp
Comment (Jinming-Hu) [reply sdk/core/azure-core/test/ut/base64.cpp:111]: Well, it's good but i don't feel it's necessary.

I don't expect the encoding/decoding impl will be changed frequently. Even it's changed and the test case fails, it's a clear message to the developer that the last change he's made is problematic. He could find some other way to debug.
So I think improving this piece of test code is some kind of investment that we can hardly benefit from.

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/base64.cpp:111]: I am not a fan of using random/non-deterministic test data for this exact reason. The input length in some cases is 10K characters, so printing to a console might not be the best.

That said, gtest's `EXPECT_EQ` is showing us the data since that is what we are comparing the round-tripped transformed result against. We pass-in `data` as a parameter. Therefore, we don't need to print that to console ourselves:
```text
..\..\..\sdk\core\azure-core\test\ut\base64.cpp(112): error: Expected equality of these values:
      Base64Decode(Base64Encode(wrong))
        Which is: { '\x1' (1), '\x2' (2), '\x3' (3) }
      data
        Which is: { '\xC0' (192), '\x82' (130), '\x89' (137), '{' (123, 0x7B), '\xEE' (238), '\xA5' (165), '\xAB' (171), '\x99' (153), 'A' (65, 0x41), '6' (54, 0x36) }
```

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:14]: never mind, I didn't think it through.

**Comment by vhvb1989** [comment]
> @Jinming-Hu, @vhvb1989 - looking for a review and approval before next week :)

For Azure Core size this is fine already.
Just wait for Storage approval and the auto-merge label should be enough

**Review by Jinming-Hu — APPROVED** [review]

**Comment by Jinming-Hu** [comment]
@msftbot hey, why don't you merge this PR?

**Pull request merged**
